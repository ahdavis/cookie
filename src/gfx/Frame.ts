/*
 * Frame.ts
 * Defines a class that represents a rendering frame
 * Created on 11/26/2019
 * Created by Andrew Davis
 *
 * Copyright (C) 2019  Andrew Davis
 *
 * This program is free software: you can redistribute it and/or modify   
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//imports
import { PixelValue } from './PixelValue';
import { Constants } from '../util/Constants';

/**
 * A rendering frame
 */
export class Frame {
	//field
	/**
	 * The pixels that make up the frame
	 */
	private _pixels: PixelValue[][];

	//methods
	
	/**
	 * Constructs a new `Frame` instance
	 */
	constructor() {
		//allocate the row array
		this._pixels = new Array<Array<PixelValue>>(
						Constants.WIN_WIDTH);

		//allocate the column arrays
		for(let x = 0; x < Constants.WIN_WIDTH; x++) {
			this._pixels[x] = new Array<PixelValue>(
						Constants.WIN_HEIGHT);
		}

		//and initialize the entire array
		for(let x = 0; x < Constants.WIN_WIDTH; x++) {
			for(let y = 0; y < Constants.WIN_HEIGHT; y++) {
				this._pixels[x][y] = PixelValue.OFF;
			}
		}
	}

	/**
	 * Gets the value of the pixel at given x and y coordinates
	 *
	 * @param x The x-coordinate to check
	 * @param y The y-coordinate to check
	 *
	 * @returns The value of the pixel at the given coordinates
	 */
	public valueAtCoords(x: number, y: number): PixelValue {
		//validate the coordinates
		if((x < 0) || (x >= Constants.WIN_WIDTH)) {
			return PixelValue.OFF;
		}
		if((y < 0) || (y >= Constants.WIN_HEIGHT)) {
			return PixelValue.OFF;
		}

		//and return the value
		return this._pixels[x][y];
	}

	/**
	 * Toggles the pixel at given coordinates
	 *
	 * @param x The x-coord of the pixel to toggle
	 * @param y The y-coord of the pixel to toggle
	 *
	 * @returns Whether the pixel was toggled from on to off
	 */
	public toggleAtCoords(x: number, y: number): boolean {
		//validate the coordinates
		if((x < 0) || (x >= Constants.WIN_WIDTH)) {
			return false;
		}
		if((y < 0) || (y >= Constants.WIN_HEIGHT)) {
			return false;
		}

		//and toggle the pixel
		if(this._pixels[x][y] == PixelValue.OFF) {
			this._pixels[x][y] = PixelValue.ON;
			return false;
		} else {
			this._pixels[x][y] = PixelValue.OFF;
			return true;
		}
	}

	/**
	 * Clears the `Frame`
	 */
	public clear(): void {
		//loop and turn off each pixel
		for(let x = 0; x < Constants.WIN_WIDTH; x++) {
			for(let y = 0; y < Constants.WIN_HEIGHT; y++) {
				this._pixels[x][y] = PixelValue.OFF;
			}
		}
	}
}

//end of file
