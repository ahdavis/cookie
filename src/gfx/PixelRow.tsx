/*
 * PixelRow.tsx
 * Defines a React component that represents a pixel row
 * Created on 11/27/2019
 * Created by Andrew Davis
 *
 * Copyright (C) 2019  Andrew Davis
 *
 * This program is free software: you can redistribute it and/or modify   
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//imports
import * as React from 'react';
import { Pixel } from './Pixel';
import { PixelRowProps } from '../props/PixelRowProps';
import { PixelRowState } from '../state/PixelRowState';
import { PixelValue } from './PixelValue';
import { Constants } from '../util/Constants';

/**
 * A row of emulator [[Pixel]]s
 */
export class PixelRow extends React.Component<PixelRowProps, PixelRowState>
{
	//no fields

	//methods

	/**
	 * Constructs a new `PixelRow` instance
	 *
	 * @param props The properties of the `PixelRow`
	 */
	constructor(props: PixelRowProps) {
		super(props); //call the superclass constructor
	}

	/**
	 * Renders the `PixelRow`
	 *
	 * @returns The React DOM for the `PixelRow`
	 */
	public render(): React.ReactNode {
		return (
			<tr>{this.assembleDOM()}</tr>
		);
	}

	/**
	 * Assembles the React DOM for the `PixelRow`
	 *
	 * @returns The React DOM for the `PixelRow`
	 */
	private assembleDOM(): React.ReactNode {
		//declare the return value
		let row: React.ReactNode[] = [];

		//loop and assemble the row
		for(let i = 0; i < Constants.WIN_WIDTH; i++) {
			//get the current pixel value
			let value = this.props.frame.valueAtCoords(
						i, this.props.index);

			//assemble its key
			let key = 'pixel (' + i + ',' + this.props.index
					+ ')';

			//determine whether the pixel is on
			let on = (value === PixelValue.ON);

			//add it to the row
			row.push(<Pixel isOn={on} key={key}/>);
		}

		//and return the assembled table row
		return row;
	}
}

//end of file
