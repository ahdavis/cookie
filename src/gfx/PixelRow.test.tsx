/*
 * PixelRow.test.tsx
 * Defines unit tests for the PixelRow component
 * Created on 11/27/2019
 * Created by Andrew Davis
 *
 * Copyright (C) 2019  Andrew Davis
 *
 * This program is free software: you can redistribute it and/or modify   
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//imports
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { PixelRow } from './PixelRow';
import { Frame } from './Frame';

//start of tests
describe('PixelRow', () => {
	//this test verifies that the pixel renders without crashing
	it('Renders without crashing', () => {
	  	let frame = new Frame();
		const div = document.createElement('div');
		ReactDOM.render(<PixelRow frame={frame} index={0} />, div);
		ReactDOM.unmountComponentAtNode(div);
	});

	//this test verifies that the pixel matches its snapshot
	it('Renders to snapshot', () => {
		let frame = new Frame();
		const row = shallow(<PixelRow frame={frame} index={0} />);
		expect(row).toMatchSnapshot();
	});
});


//end of tests
