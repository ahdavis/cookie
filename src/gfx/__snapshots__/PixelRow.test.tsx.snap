// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PixelRow Renders to snapshot 1`] = `ShallowWrapper {}`;
